# Homework 3

## Exercise 2.2

1. Verify the bound of Theorem 2.4 in the three cases of Example 2.2:
  
  1. Positive rays: $\mathcal{H}$ consists of all hypothesis in one dimension of the form $h(x) = \text{sign}(x - a)$. $k = 2$
    
  2. Positive intervals: $\mathcal{H}$ consists of all hypotheses in one dimension that are positive within some interval and negative elsewhere. $k = 3$
    
  3. Convex sets: $\mathcal{H}$ consists of all hypothesis in two dimensions that are positive inside some convex set and negative elsewhere
    
  
  Note: you can use the break points you found in Exercise 2.1
  
2. Does there exist a hypothesis set for which $m_\mathcal{H}(N) = N + 2^{\lfloor N / 2 \rfloor}$ (where $\lfloor N / 2 \rfloor$ is the largest integer $\le N / 2$)?
  

### Solution:

1.1 Positive rays. We know that that $m_\mathcal{H}(N) = N + 1$. So our break point is $k = 2$.

So now what we need to write is the following,

$$
\sum_{i = 0}^{k - 1} {N \choose i} = \sum_{i = 0}^{ 1} {N \choose i} = {N! \over 0!N!} + {N! \over 1! (N-1)!} = 1 + N
$$

thanks we also know that as $k$ increases then $\sum_{i=0}^{k-1}{N \choose i}$ also increases. So the theorem is true for $k \ge 2$.

1.2 Positive intervals. We know that $m_\mathcal{H}(N) = {1 \over 2}N^2 + {1 \over 2}N + 1$. So our break point is $k = 3$.

So now we can write the following,

$$
\sum_{i = 0}^{k - 1}{N \choose i} = \sum_{i = 0}^{2}{N \choose i} = 1 + N + {N! \over 2! (N -2)!} = 1 + N + {(N-1)N \over 2}
$$$$
\sum_{i = 0}^{k - 1}{N \choose i} =  1 + N + {1 \over 2}N^2 - {1 \over 2}N = {1 \over 2}N^2 + {1 \over 2}N + 1
$$

So we know that the theorem is true for $k = 3$. Now as we know that $\sum_{i = 0}^{k - 1}{N \choose i}$ is non-decreasing so the theorem is true for $k \ge 3$.

1.3 Convex set. We know that $m_\mathcal{H}(N) = 2^N$. So the theorem is true for vacuous true.

2 It exists it is obviously connected to a bi-decision type of model (with binary-trees maybe)

## Exercise 2.5

Suppose we have a simple learning model whose growth function is $m_\mathcal{H}(N) = N + 1$, hence $d_{\text{VC}} = 1$ (Haha $k = 2$ for positive rays). Use the VC bound (2.12) to estimate the probability that $E_{\text{out}}$ will be within 0.1 of $E_{\text{in}}$ given 100 training samples. [Hint: The estimate will be ridiculous]

### Solution ?

$$
E_{\text{out}}(g) \le E_\text{in}(g) + \sqrt{{8 \over N} \ln {4m_\mathcal{H}(2N) \over \delta}}
$$$$
E_\text{out}(g) - E_\text{in}(g) \le \sqrt{{8 \over N} \ln {4 m_\mathcal{H}(2N) \over \delta}}
$$

We need that $E_\text{out}(g) - E_\text{in}(g) \le 0.1$. So this means that

$$
\sqrt{{8 \over N} \ln {4 m_\mathcal{H}(2N) \over \delta}} = 0.1
$$$$
\sqrt{{8 \over N} \left (\ln m_\mathcal{H}(2N) + \ln {4 \over \delta} \right )}  = 0.1
$$$$
\ln m_\mathcal{H}(2N) + \ln {4 \over \delta} = 0.01 {N \over 8}
$$$$
\ln 4 - \ln \delta = 0.01 {N \over 8} - \ln m_\mathcal{H}(2N)
$$$$
\ln \delta = - 0.01 {N \over 8} + \ln m_\mathcal{H}(2N) + \ln 4
$$$$
\ln \delta = 6.5646
$$$$
\delta = e^{6.5646}
$$

## Exercise 2.6

A data set has 600 examples. To properly test the perfomance of the final hypothesis, you set aside a randomly selected subset of 200 examples which are never used in the training phase; these form a test set. You use a learning model with 1000 hypotheses and select the final hypothesis $g$ based on the 400 training examples. We wish to estimate $E_{\text{out}}(g)$. We have access to two estimates: $E_{\text{in}}(g)$, the in-sample error on the 400 training examples; and, $E_{\text{test}}(g)$, the test error on the 200 test examples that were set aside.

1. Using a $5\%$ error tolerance ($\delta = 0.05$), which estimate has the higher 'error bar'?
  
2. Is there any reason why you shouldn't reserve even more examples for testing?
  
### Solution

For our $E_\text{in}$ we only need to see the VC generalization bound.

$$
E_{\text{out}}(g) \le E_\text{in}(g) + \sqrt{{8 \over N} \ln {4m_\mathcal{H}(2N) \over \delta}}
$$

with $N = 400$

$$
\sqrt{{8 \over N} ln {4m_\mathcal{H}(2N) \over \delta}} = 0,470486718
$$

and for $N = 200$ and testing we use the Hoeffding bound

$$
\delta \le 2e^{-2\epsilon^2N}
$$

$$
{ \delta \over 2} \le e^{-2\epsilon^2N}
$$

$$
\ln { \delta \over 2} \le -2\epsilon^2N
$$

$$
-{1 \over 2N} \ln { \delta \over 2} \le \epsilon^2
$$

$$
\sqrt{ -{1 \over 2N} \ln { \delta \over 2} } \le \epsilon
$$

$$
\epsilon = 0,096032279
$$

2 The obvious reason we can see from the bounds, the general VC bound been losser when you decrease the number of samples has a more extreme consequence, so you are less sure of the model you selected but more sure of the error when testing. Just caring about the test without caring about the learning process will just give non-good hypotheses.

## Problem 2.1

In Equation (2.1), set $\delta = 0.03$ and let

$$
\epsilon(M, N, \delta)  = \sqrt{{1 \over 2N} \ln {2M \over \delta}}.
$$

1. For $M = 1$, how many examples do we need to make $\epsilon \le 0.05$?
  
2. For $M = 100$, how many examples do we need to make $\epsilon \le 0.05$?
  
3. For $M = 10000$, how many examples do we need to make $\epsilon \le 0.05$?
  

### Solution

1

$$
0.05 \ge {1 \over 2N} \ln {2 \over 0.03}.
$$$$
N \ge {1 \over 0.1} \ln {2 \over 0.03}
$$$$
N \ge 41.997
$$

We need 42 examples or more.

2.

$$
0.05 \ge {1 \over 2N} \ln {2 \over 0.03}.
$$$$
N \ge {1 \over 0.1} \ln {200 \over 0.03}
$$$$
N \ge 88,048752639
$$

We need 89 examples or more

3

$$
0.05 \ge {1 \over 2N} \ln {20000 \over 0.03}.
$$$$
N \ge {1 \over 0.1} \ln {20000 \over 0.03}
$$$$
N \ge 134,100454499
$$

We need 135 examples or more

## Problem 2.12

For an $\mathcal{H}$ with $d_{\text{VC}} = 10$, what sample size do you need (as prescribed by the generalization bound) to have a $95\%$ confidence that your generalization error is at most $0.05$?

We use the equation $(2.13)$

$$
N \ge {8 \over \epsilon^2} \ln \left ( {4((2N)^{d_\text{VC}} + 1) \over \delta} \right )
$$

In [3]:
import math

N_0 = 100
N_1 = None
while True:
    N_1 = 8/(0.05)**2 * math.log((4 * ((2 * N_0)**10 + 1))/0.95)
    N_1 = math.ceil(N_1)
    if N_1 == N_0:
        break
    N_0 = N_1
print(N_1)

442810


Well it seems that the we need a sample size of $442810$

## Problem 2.16

In this problem, we will consider $\mathcal{X} = \mathbb{R}$. That is, $\mathbf{x} = x$ is a one-dimensional variable. For a hypothesis set

$$
\mathcal{H} = \left \lbrace h_{\mathbf{c}} | h_{\mathbf{c}}(x) = \text{sign} \left ( \sum_{i = 0}^{D} c_ix^i \right ) \right \rbrace
$$

prove that the VC dimension of $\mathcal{H}$ is exactly $(D + 1)$ by showing that

1. There are $(D + 1)$ points which are shattered by $\mathcal{H}$.
  
2. There are no $(D + 2)$ points which are shattered by $\mathcal{H}$.
  

### Solution

1 To prove that there are $(D+1)$ points shattered by $\mathcal{H}$ we only need to use the values of $\mathbf{c}$. We use the values of $+1, -1, 0$. For separating the values.

For example $x = 1$ and $x = 2$, we need to separate this to values then we use $c_1 = 1$ and $c_2 = -1$. Also we can used the other value $c_0$ to separate the rest of the entries (outside the examples), so we can separate $(D + 1)$.

2 We notice that $\sum_{i=0}^{D} c_i x^i$ is a polinomial of degree $D$, but also we can imagine the roots as opportunities to change the classification on the $\mathbb{R}$, so each time we encounter a root the function can go up again (imagine a parabola) or it can go down (a sine wave is the simplest example that comes to mind).

So you must start with an option $(+1, -1)$ then you have $D$ possible changes to shattered $(D + 2)$ you would need $D + 1$ possible changes.